packages/flutter/lib/src/widgets/selection_arena.dart

```diff
@@ -0,0 +1,190 @@
+import 'dart:math' as math;
+
+import 'package:flutter/gestures.dart';
+import 'package:flutter/rendering.dart';
+
+import '../../widgets.dart';
+
+class SelectionArea extends StatefulWidget {
+  const SelectionArea({required this.child, Key? key}) : super(key: key);
+
+  final Widget child;
+
+  @override
+  _SelectionAreaState createState() => _SelectionAreaState();
+
+  static SelectionRegistrant? of(BuildContext context) {
+    return context.findAncestorStateOfType<_SelectionAreaState>();
+  }
+}
+
+class _SelectionState {
```

Comment:
nit: *State is a not an ideal name for this given that State objects usually belong to StatefulWidgets.